Src/AmrCore/AMReX_ErrorList.cpp

```diff
@@ -228,4 +228,190 @@ operator << (std::ostream&    os,
     return os;
 }
 
+  static
+  void
+  AMRErrorTag_GRAD(const Box&                bx,
+                   Array4<const Real> const& dat,
+                   Array4<char> const&       tag,
+                   Real                      threshold,
+                   char                      tagval)
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      auto ax = amrex::Math::abs(dat(i+1,j,k) - dat(i,j,k));
+      ax = amrex::max(ax,amrex::Math::abs(dat(i,j,k) - dat(i-1,j,k)));
+#if AMREX_SPACEDIM == 1
+      if (ax >= threshold) tag(i,j,k) = tagval;
+#else
+      auto ay = amrex::Math::abs(dat(i,j+1,k) - dat(i,j,k));
+      ay = amrex::max(ay,amrex::Math::abs(dat(i,j,k) - dat(i,j-1,k)));
+#if AMREX_SPACEDIM > 2
+      auto az = amrex::Math::abs(dat(i,j,k+1) - dat(i,j,k));
+      az = amrex::max(az,amrex::Math::abs(dat(i,j,k) - dat(i,j,k-1)));
+#endif
+      if (amrex::max(D_DECL(ax,ay,az)) >= threshold) {
+        tag(i,j,k) = tagval;
+      }
+#endif
+    });
+  }
+
+  int
+  AMRErrorTag::SetNGrow () const noexcept
+  {
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(m_test != USER, "Do not call SetNGrow with USER test");
+    static std::map<TEST,int> ng = { {GRAD,1}, {LESS,0}, {GREATER,0}, {VORT,0}, {BOX,0} };
+    return ng[m_test];
+  }
+  
+  static
+  void
+  AMRErrorTag_LESS(const Box&                bx,
+                   Array4<const Real> const& dat,
+                   Array4<char> const&       tag,
+                   Real                      threshold,
+                   char                      tagval) noexcept
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      if (dat(i,j,k) <= threshold) {
+        tag(i,j,k) = tagval;
+      }
+    });
+  }
+
+  static
+  void
+  AMRErrorTag_GREATER(const Box&                bx,
+                      Array4<const Real> const& dat,
+                      Array4<char> const&       tag,
+                      Real                      threshold,
+                      char                      tagval) noexcept
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      if (dat(i,j,k) >= threshold) {
+        tag(i,j,k) = tagval;
+      }
+    });
+  }
+
+  static
+  void
+  AMRErrorTag_BOX(const Box&          bx,
+                  Array4<char> const& tag,
+                  const RealBox&      tag_rb,
+                  const Geometry&     geom,
+                  char                tagval) noexcept
+  {
+    auto plo = geom.ProbLoArray();
+    auto dx  = geom.CellSizeArray();
+    class RealBox trb(bx,dx.data(),plo.data());
+    if (tag_rb.intersects(trb))
+    {
+      amrex::ParallelFor(bx,
+      [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+      {
+        GpuArray<Real,AMREX_SPACEDIM> pt = {D_DECL(plo[0]+(i+0.5)*dx[0],
```

Comment:
@drummerdoc @WeiqunZhang Getting a AppleClang (SP) and DPC++ (DP) compile error on this of the form:
```
Src/AmrCore/AMReX_ErrorList.cpp:318:52: error: non-constant-expression cannot be narrowed from type 'double' to 'float' in initializer list [-Wc++11-narrowing]
```

in WarpX.

Reply:
maybe we can cast the `i`, `j`, `k` with `Real(i)` etc. and add `_rt` to `+ 0.5` as `+ 0.5_rt` to fix it.

Reply:
Let's try this: #1180